packages/flutter/lib/src/material/list_tile.dart

```diff
@@ -630,57 +632,69 @@ class ListTile extends StatelessWidget {
     return dense ?? tileTheme.dense ?? theme.listTileTheme.dense ?? false;
   }
 
-  TextStyle _titleTextStyle(ThemeData theme, ListTileThemeData tileTheme) {
-    final TextStyle textStyle;
-    switch(style ?? tileTheme.style ?? theme.listTileTheme.style ?? ListTileStyle.list) {
-      case ListTileStyle.drawer:
-        textStyle = theme.useMaterial3 ? theme.textTheme.bodyMedium! : theme.textTheme.bodyLarge!;
-        break;
-      case ListTileStyle.list:
-        textStyle = theme.useMaterial3 ? theme.textTheme.titleMedium! : theme.textTheme.titleMedium!;
-        break;
+  TextStyle _titleTextStyle(ThemeData theme, ListTileThemeData tileTheme, ListTileThemeData defaults) {
+    final TextStyle textStyle = titleTextStyle
+      ?? tileTheme.titleTextStyle
+      ?? defaults.titleTextStyle!;
+    if (theme.useMaterial3) {
+      return textStyle;
     }
     final Color? color = _textColor(theme, tileTheme, textStyle.color);
     return _isDenseLayout(theme, tileTheme)
       ? textStyle.copyWith(fontSize: 13.0, color: color)
       : textStyle.copyWith(color: color);
   }
 
-  TextStyle _subtitleTextStyle(ThemeData theme, ListTileThemeData tileTheme) {
-    final TextStyle textStyle = theme.useMaterial3 ? theme.textTheme.bodyMedium! : theme.textTheme.bodyMedium!;
-    final Color? color = _textColor(
-      theme,
-      tileTheme,
-      theme.useMaterial3 ? theme.textTheme.bodySmall!.color : theme.textTheme.bodySmall!.color,
-    );
+  TextStyle _subtitleTextStyle(ThemeData theme, ListTileThemeData tileTheme, ListTileThemeData defaults) {
+    final TextStyle textStyle = subtitleTextStyle
+      ?? tileTheme.subtitleTextStyle
+      ?? defaults.subtitleTextStyle!;
+    if (theme.useMaterial3) {
+      return textStyle;
+    }
+    final Color? color = _textColor(theme, tileTheme, theme.textTheme.bodySmall!.color);
     return _isDenseLayout(theme, tileTheme)
       ? textStyle.copyWith(color: color, fontSize: 12.0)
       : textStyle.copyWith(color: color);
   }
 
-  TextStyle _trailingAndLeadingTextStyle(ThemeData theme, ListTileThemeData tileTheme) {
-    final TextStyle textStyle = theme.useMaterial3 ? theme.textTheme.bodyMedium! : theme.textTheme.bodyMedium!;
+  TextStyle _trailingAndLeadingTextStyle(ThemeData theme, ListTileThemeData tileTheme, ListTileThemeData defaults) {
+    final TextStyle textStyle = trailingAndLeadingTextStyle
+      ?? tileTheme.trailingAndLeadingTextStyle
+      ?? defaults.trailingAndLeadingTextStyle!;
+    if (theme.useMaterial3) {
+      return textStyle;
+    }
     final Color? color = _textColor(theme, tileTheme, textStyle.color);
     return textStyle.copyWith(color: color);
```

Comment:
Is there no way to make this work for M2 with just the defaults? Or is there processing of the color that the M2 defaults couldn't account for?

Reply:
These properties (`textColor` and `iconColor`) override defaults based on `states`. However, I've refactored this code and got rid of a bunch of methods that were calculating disabled, selected states, and added a generic `_resolveState` method. Please take a look, this makes the code consistent with other M3 widgets.

Before:
```dart
  Color? _textColor(ThemeData theme, ListTileThemeData tileTheme, Color? defaultColor) {
    if (!enabled) {
      return theme.disabledColor;
    }

    if (selected) {
      return selectedColor ?? tileTheme.selectedColor ?? theme.listTileTheme.selectedColor ?? theme.colorScheme.primary;
    }

    return textColor ?? tileTheme.textColor ?? theme.listTileTheme.textColor ?? defaultColor;
  }
    // ...
  TextStyle _trailingAndLeadingTextStyle(ThemeData theme, ListTileThemeData tileTheme) {
    final TextStyle textStyle = theme.useMaterial3 ? theme.textTheme.bodyMedium! : theme.textTheme.bodyMedium!;
    final Color? color = _textColor(theme, tileTheme, textStyle.color);
    return textStyle.copyWith(color: color);
  }
    // ...
    TextStyle? leadingAndTrailingTextStyle;
    if (leading != null || trailing != null) {
      leadingAndTrailingTextStyle = _trailingAndLeadingTextStyle(theme, tileTheme);
    }
```

After it looks something like this:
```dart
    final Color? effectiveTextColor = textColor
      ?? tileTheme.textColor
      ?? theme.listTileTheme.textColor;
    
    final Color effectiveSelectedColor = selectedColor
      ?? tileTheme.selectedColor
      ?? theme.listTileTheme.selectedColor
      ?? defaults.selectedColor!;
      //...
      leadingAndTrailingStyle = leadingAndTrailingTextStyle
        ?? tileTheme.leadingAndTrailingTextStyle
        ?? defaults.leadingAndTrailingTextStyle!;
      final Color? leadingAndTrailingTextColor = _resolveState(
        value: effectiveTextColor,
        selectedValue: effectiveSelectedColor,
        disabledValue: theme.disabledColor,
        states: states,
      );
      leadingAndTrailingStyle = leadingAndTrailingStyle.copyWith(color: leadingAndTrailingTextColor);
```

Reply:
If we're going to add support for specifying Color properties with MaterialStateColors and if we're going to also specify the M2/M3 defaults that way, AND we're going to preserve backwards compatibility with properties like hoverColor and selectedColor, then this is going to get a bit complicated.

When a color property can be a MaterialStateColor _or_ a Color you need to resolve it with `MaterialStatePropert.resolveAs<Color?>(value, states)`. When the value of such a property is defined by a widget or its theme or its default then you must write:

```dart
Color effectiveColor = MaterialStatePropert.resolveAs<Color?>(widgetValue, states)
  ?? MaterialStatePropert.resolveAs<Color?>(themeValue, states)
  ?? MaterialStatePropert.resolveAs<Color>(defaultValue, states);
```

That way the wiget or the theme can specify a MaterialStateColor that only resolves to non-null for the states they want to override.

If we also want to support old properties that define a color for just one state, like hoverColor, then you'll need to weave them in using something like _resolveState. I think it would be cleaner to just use `MaterialStateProperty<Color?>` objects for these cases:

```dart
Color effectiveColor = MaterialStatePropert.resolveAs<Color?>(widgetValue, states)
  ?? _IndividualOverrides(hover: widgetHoverColor, focus: widgetFocusColor, disabled: widgetDisabledColor)
  ?? MaterialStatePropert.resolveAs<Color?>(themeValue, states)
  ?? _IndividualOverrides(hover: themeHoverColor, focused: themeFocusColor, disabled: themeDisabledColor)
  ?? MaterialStatePropert.resolveAs<Color>(defaultValue, states);
```

Where _IndividualOverrides is a class that extends `MaterialStateProperty<Color?>` and has a resolve method that DRT with the specified state-specific parameters.

Reply:
Thank you so much for taking the time to explain this idea!

This makes sense, I did something similar for Slider a few weeks ago, `SliderTheme.overlayColor` (Color type) now uses `MaterialStateProperty.resolveAs` and the user can provide states for the existing `Color?` type `SliderTheme.overlay` property.

`_IndividualOverrides` seems suitable for hover, focus, and overlay, but `_resolveState` checks for `selected` states to apply selected and disabled color. These are different sets of colors. Hover, focus, soon to be added overlay are applied to `InkWell` but `_resolveState` color is applied to text styles and icons. I'm figuring this out to get rid of  `_resolveState` that I added in this PR.

Reply:
Updated the PR with `MaterialStateProperty.resolveAs` (except `tileColor` & `selectedTileColor`)  and added `_IndividualOverrides` for backward compatibility.  Also added 4 more new tests to check for `MaterialStateProperty.resolveAs` in the widget and theme classes.

> That way the widget or the theme can specify a MaterialStateColor that only resolves to non-null for the states they want to override.

This allows updating `ListTile.iconColor` & `ListTile.iconColor` using `MaterialStateColor`, instead of using both regular and selected color as suggested above. 

In part 2 PR of this migration, I will add M3 overlay and new padding, and alignments. 

(other list tile widgets depend on `tileColor & `selectedTileColor` and they expect to get the selected color without setting `ListTile.selected` to true, I'll tackle this in a separate PR as it may require updating and adding tests to other list tile widgets(